python/ml4ir/model/ranking_model.py

```diff
@@ -721,12 +732,47 @@ def _get_dense_feature(inputs, feature, shape=(1,)):
                     inputs, feature_node_name, shape=(self.max_num_records, 1)
                 )
                 if feature_info["trainable"]:
-                    ranking_features.append(tf.cast(dense_feature, tf.float32))
+                    dense_feature = tf.expand_dims(tf.cast(dense_feature, tf.float32), axis=-1)
+                    ranking_features.append(dense_feature)
                 else:
                     metadata_features[feature_node_name] = tf.cast(dense_feature, tf.float32)
             elif feature_layer_info["type"] == FeatureTypeKey.STRING:
-                # TODO: Add embedding layer here
-                pass
+                if feature_info["trainable"]:
+                    if feature_layer_info["embedding_type"] == EmbeddingTypeKey.BILSTM:
+                        """
+                        NOTE: The input shape MUST be
+                        [batch_size, time_steps, feature]
+                        """
+                        input_feature = tf.cast(inputs[feature_node_name], tf.uint8)
+                        input_feature = tf.reshape(
+                            input_feature, [-1, feature_layer_info["max_length"]]
+                        )
+                        input_feature = tf.one_hot(input_feature, depth=256)
+
+                        dense_feature = layers.Bidirectional(
+                            layers.LSTM(
+                                int(feature_layer_info["embedding_size"] / 2),
+                                return_sequences=False,
+                            ),
+                            merge_mode="concat",
+                        )(input_feature)
+                        if feature_info["tfrecord_type"] == TFRecordTypeKey.CONTEXT:
+                            # If feature is a context feature then tile it for all records
+                            dense_feature = tf.expand_dims(dense_feature, axis=1)
+                            dense_feature = kbackend.repeat_elements(
```

Comment:
Just an FYI. Doing first the BiLSTM and then the tile

Reply:
yep.  but probably extract this out to another method.

Reply:
Yes, I am planning on doing that soon. Letting some models train for now.